src/crypto.rs

```diff
@@ -18,24 +18,103 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-use openssl::{
-    ec::{EcGroup, EcKey},
-    error::ErrorStack,
-    hash::MessageDigest,
-    nid::Nid,
-    pkey::PKey,
-    x509::{X509NameBuilder, X509ReqBuilder},
+use std::str::FromStr;
+
+use p384::{
+    ecdsa::{DerSignature, SigningKey},
+    pkcs8::{EncodePrivateKey, LineEnding},
+    SecretKey,
+};
+use rustls::PrivateKey;
+use x509_cert::{
+    builder::{Builder, RequestBuilder},
+    der::EncodePem,
+    name::Name,
 };
 
-pub struct Bundle(pub Vec<u8>, pub Vec<u8>);
+/// Errors that can occur while generating the Certificate and CSR.
+#[non_exhaustive]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[cfg(feature = "openssl")]
+    #[error("Openssl error")]
+    Openssl(#[from] openssl::error::ErrorStack),
+
+    #[error("Invalid COMMONNAME")]
+    InvalidCn(#[from] x509_cert::der::Error),
+
+    #[error("Failed to create Certificate and CSR")]
+    Certificate(#[from] x509_cert::builder::Error),
+
+    #[error("Failed to encode key to PKCS8 pem")]
+    Pem(#[from] p384::pkcs8::Error),
+
+    #[error("Invalid UTF-8 encoded PEM")]
+    Utf8(#[from] std::string::FromUtf8Error),
+}
+
+#[cfg(windows)]
+const LINE_ENDING: LineEnding = LineEnding::CRLF;
+
+#[cfg(not(windows))]
+const LINE_ENDING: LineEnding = LineEnding::LF;
+
+/// Generate a Certificate and CSR bundle in PEM format.
+#[derive(Debug)]
+pub struct Bundle {
+    pub private_key: PrivateKey,
+    /// PEM encoded CSR
+    pub csr: String,
+}
 
 impl Bundle {
-    pub fn new(cn: &str) -> Result<Bundle, ErrorStack> {
+    pub fn new(realm: &str, device_id: &str) -> Result<Bundle, Error> {
```

Comment:
Why keep both implementations available at all times?
The implementation can be simpler if only one of them is active.
The two functions could be renamed to have the same name (`generate_key`).
The directive `#[cfg(not(feature = "openssl"))]` can be used to compile only the version that is selected.
If a user wants to use `openssl` it will most likely use it always, so no need to provide both methods.